Mathlib/Algebra/Category/ModuleCat/Basic.lean

```diff
@@ -88,27 +71,121 @@ instance : CoeSort (ModuleCat.{v} R) (Type v) :=
+  ⟨X⟩
+
+lemma coe_of (X : Type v) [Ring X] [Module R X] : (of R X : Type v) = X :=
+  rfl
+
+-- Ensure the roundtrips are reducibly defeq (so tactics like `rw` can see through them).
+example (X : Type v) [Ring X] [Module R X] : (of R X : Type v) = X := by with_reducible rfl
+example (M : ModuleCat.{v} R) : of R M = M := by with_reducible rfl
+
+variable {R} in
+/-- The type of morphisms in `ModuleCat R`. -/
+@[ext]
+structure Hom (M N : ModuleCat.{v} R) where
+  private mk ::
+  /-- The underlying linear map. -/
+  hom : M →ₗ[R] N
+
 instance moduleCategory : Category.{v, max (v+1) u} (ModuleCat.{v} R) where
-  Hom M N := M →ₗ[R] N
-  id _ := LinearMap.id
-  comp f g := g.comp f
-  id_comp _ := LinearMap.id_comp _
-  comp_id _ := LinearMap.comp_id _
-  assoc f g h := LinearMap.comp_assoc (f := f) (g := g) (h := h)
+  Hom M N := Hom M N
+  id _ := ⟨LinearMap.id⟩
+  comp f g := ⟨g.hom.comp f.hom⟩
+
+instance {M N : ModuleCat.{v} R} : CoeFun (M ⟶ N) (fun _ ↦ M → N) where
+  coe f := f.hom
+
+section
+
+variable {R}
+
+@[simp]
+lemma hom_id {M : ModuleCat.{v} R} : (𝟙 M : M ⟶ M).hom = LinearMap.id := rfl
+
+/- Provided for rewriting. -/
+lemma id_apply (M : ModuleCat.{v} R) (x : M) :
+    (𝟙 M : M ⟶ M) x = x := by simp
+
+@[simp]
+lemma hom_comp {M N O : ModuleCat.{v} R} (f : M ⟶ N) (g : N ⟶ O) :
+    (f ≫ g).hom = g.hom.comp f.hom := rfl
+
+/- Provided for rewriting. -/
+lemma comp_apply {M N O : ModuleCat.{v} R} (f : M ⟶ N) (g : N ⟶ O) (x : M) :
+    (f ≫ g) x = g (f x) := by simp
+
+@[ext]
+lemma hom_ext {M N : ModuleCat.{v} R} {f g : M ⟶ N} (hf : f.hom = g.hom) : f = g :=
+  Hom.ext hf
+
+lemma hom_bijective {M N : ModuleCat.{v} R} :
+    Function.Bijective (Hom.hom : (M ⟶ N) → (M →ₗ[R] N)) where
+  left f g h := by cases f; cases g; simpa using h
+  right f := ⟨⟨f⟩, rfl⟩
+
+/-- Typecheck a `LinearMap` as a morphism in `ModuleCat R`. -/
+abbrev asHom {X Y : Type v} [AddCommGroup X] [Module R X] [AddCommGroup Y] [Module R Y]
+    (f : X →ₗ[R] Y) : of R X ⟶ of R Y :=
+  ⟨f⟩
+
+@[deprecated (since := "2024-10-06")] alias ofHom := ModuleCat.asHom
+
+lemma hom_asHom {X Y : Type v} [AddCommGroup X] [Module R X] [AddCommGroup Y]
+    [Module R Y] (f : X →ₗ[R] Y) : (asHom f).hom = f := rfl
 
-instance {M N : ModuleCat.{v} R} : FunLike (M ⟶ N) M N :=
-  LinearMap.instFunLike
+@[simp]
+lemma asHom_hom {M N : ModuleCat.{v} R} (f : M ⟶ N) :
+    asHom (Hom.hom f) = f := rfl
+
+@[simp]
+lemma asHom_id {M : Type v} [AddCommGroup M] [Module R M] : asHom LinearMap.id = 𝟙 (of R M) := rfl
+
+@[simp]
+lemma asHom_comp {M N O : Type v} [AddCommGroup M] [AddCommGroup N] [AddCommGroup O] [Module R M]
+    [Module R N] [Module R O] (f : M →ₗ[R] N) (g : N →ₗ[R] O) :
+    asHom (g.comp f) = asHom f ≫ asHom g :=
+  rfl
 
-instance {M N : ModuleCat.{v} R} : LinearMapClass (M ⟶ N) R M N :=
-  LinearMap.semilinearMapClass
+lemma asHom_apply {M N : Type v} [AddCommGroup M] [AddCommGroup N] [Module R M] [Module R N]
+    (f : M →ₗ[R] N) (x : M) : asHom f x = f x := rfl
```

Comment:
Similarly here, either `@[simp]` or explain why it's not a good idea.

Reply:
I thought this would be the same issue, but actually it would be a good `@[simp]` lemma!

Reply:
This is solvable by `simp`, so I don't think we should add `@[simp]` here.